docs/docs/canonicai/chapter3.md

```diff
@@ -0,0 +1,339 @@
+
+* **NLG (Natural Language Generation)**: Natural language generation (NLG) is the use of artificial intelligence (AI) programming to produce written or spoken narratives from a data set. **What does NLG do?** If you look above at the diagram, you would see after the AI understand what the user is trying to say using NLU it will then take those results and auto generate a suitable response to forward to the user and that is the purpose of the NLG.
+
+### **What do each node abilities do?**
+
+* **classifiy_intent**: Figures out the user intention in any given question.
+* **extract_entities**: Extract words of interests from the question
+* **gen_response**: Generate a suitable response for the user.
+* **nlu**: Process incoming request through NLU engines using intent classification or entity extraction AI models.
+* **nlg**: Construct natural language response
+
+
+### **Inheritance**
+**What is inheritance?**
+ It is a mechanism where you can to derive a class from another class for a hierarchy of classes that share a set of attributes and methods.
+ 
+ **Node Inheritance**
+``` 
+node vehicle {
+    has plate_number;
+    
+    can drive {
+        report 'yes';
+    }
+}
+
+node car:vehicle {
+    has plate_number = "RAC001";
+}
+
+node bus:vehicle {
+    has plate_number = "SUB002";
+}
+```
+In the above code snippet. This is a very basic example of inheritance. We created a node named car and bus but since they are related and simply share the same property and node ability we created a node named vehicle and inherit its attributes. Imagine life without inheritance, the code will look like this.
+``` 
+node car {
+    has plate_number = "RAC001";
+    
+    can drive {
+        report 'yes';
+    }
+}
+
+node bus {
+    has plate_number = "SUB002";
+    
+    can drive {
+        report 'yes';
+    }
+}
+```
+In this case it will seem that writing it this way will be more effient, however what if each vehicles have more functions like blow_horn, accelerate, breaks, park and etc or they are more vehicles like truck, scooter, and etc you can imagine how big this code base will be and that's why it's important to have inheritance in our code base and in this program we use inheritance to increase the efficiency of the entire application.
+
+So, lets use what's in the code base and see how we implemented it. ```Do not worry about what each nodes and state works, that will be explained in detailed in the next section, just look at how it is implemented.```
+```
+node cai_state {
```

Comment:
Let's break this down into each individual node ability and explain what they do.

Reply:
It's right after that section where we explain collect_info state, confirmation state and etc